coderd/insights.go

```diff
@@ -43,3 +51,329 @@ func (api *API) deploymentDAUs(rw http.ResponseWriter, r *http.Request) {
 	}
 	httpapi.Write(ctx, rw, http.StatusOK, resp)
 }
+
+// @Summary Get insights about user latency
+// @ID get-insights-about-user-latency
+// @Security CoderSessionToken
+// @Produce json
+// @Tags Insights
+// @Success 200 {object} codersdk.UserLatencyInsightsResponse
+// @Router /insights/user-latency [get]
+func (api *API) insightsUserLatency(rw http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	if !api.Authorize(r, rbac.ActionRead, rbac.ResourceDeploymentValues) {
+		httpapi.Forbidden(rw)
+		return
+	}
+
+	// TODO(mafredri): Client or deployment timezone?
+	// Example:
+	// - I want data from Monday - Friday
+	// - I'm UTC+3 and the deployment is UTC+0
+	// - Do we select Monday - Friday in UTC+0 or UTC+3?
+	// - Considering users can be in different timezones, perhaps this should be per-user (but we don't keep track of user timezones).
+	p := httpapi.NewQueryParamParser().
+		Required("start_time").
+		Required("end_time")
+	vals := r.URL.Query()
+	var (
+		startTime   = p.Time3339Nano(vals, time.Time{}, "start_time")
+		endTime     = p.Time3339Nano(vals, time.Time{}, "end_time")
+		templateIDs = p.UUIDs(vals, []uuid.UUID{}, "template_ids")
+	)
+	p.ErrorExcessParams(vals)
+	if len(p.Errors) > 0 {
+		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+			Message:     "Query parameters have invalid values.",
+			Validations: p.Errors,
+		})
+		return
+	}
+
+	if !verifyInsightsStartAndEndTime(ctx, rw, startTime, endTime) {
+		return
+	}
+
+	// Should we verify all template IDs exist, or just return no rows?
+	// _, err := api.Database.GetTemplatesWithFilter(ctx, database.GetTemplatesWithFilterParams{
+	// 	IDs: templateIDs,
+	// })
+
```

Comment:
principle of least surprise: if we request data for templates (A, B, C) and template B no longer exists, just return data for templates (A, C).